core/src/query.rs

```diff
@@ -13,47 +17,92 @@ pub struct MongoQuery {
+
+                // 1. Run the $sql aggregation to get the result set cursor.
+                let pipeline = vec![doc! {"$sql": {
+                    "dialect": "mongosql",
+                    "format": "odbc",
+                    "statement": query,
+                }}];
+
+                let options = query_timeout.map(|i| {
+                    AggregateOptions::builder()
+                        .max_time(Duration::from_millis(i as u64)) // TODO: should timeout come from client if present?
+                        .build()
+                });
+
+                let cursor = db.aggregate(pipeline, options)?;
+
+                // 2. Run the sqlGetResultSchema command to get the result set
+                // metadata. Sort the column metadata alphabetically by column
+                // name.
+                let get_result_schema_cmd =
+                    doc! {"sqlGetResultSchema": 1, "query": query, "schemaVersion": 1};
+
+                let get_result_schema_response: SqlGetResultSchemaResponse =
+                    bson::from_document(db.run_command(get_result_schema_cmd, None)?)
+                        .map_err(Error::BsonDeserialization)?;
+
+                let metadata = get_result_schema_response.process_metadata()?;
+
+                Ok(MongoQuery {
+                    resultset_cursor: cursor,
+                    resultset_metadata: metadata,
+                })
+            }
+        }
     }
 
     // Return the number of fields/columns in the resultset
     fn _get_col_count(&self) -> u32 {
-        unimplemented!()
+        self.resultset_metadata.len() as u32
     }
 
     // Get the metadata for the column with the given index.
-    fn _get_col_metadata(&self, _col_index: u16) -> Result<MongoColMetadata> {
-        unimplemented!()
+    fn _get_col_metadata(&self, col_index: u16) -> Result<&MongoColMetadata> {
+        self.resultset_metadata
+            .get(col_index as usize)
+            .map_or(Err(Error::ColIndexOutOfBounds(col_index)), |md| Ok(md))
     }
 }
 
 impl MongoStatement for MongoQuery {
     // Move the cursor to the next document and update the current row.
     // Return true if moving was successful, false otherwise.
     fn next(&mut self) -> Result<bool> {
-        unimplemented!()
+        self.resultset_cursor.advance().map_err(Error::MongoError)
     }
 
     // Get the BSON value for the cell at the given colIndex on the current row.
     // Fails if the first row as not been retrieved (next must be called at least once before getValue).
-    fn get_value(&self, _col_index: u16) -> Result<Option<&Bson>> {
-        unimplemented!()
+    fn get_value(&self, col_index: u16) -> Result<Option<&Bson>> {
+        let md = self._get_col_metadata(col_index)?;
+        let datasource = self
+            .resultset_cursor
+            .deserialize_current()?
+            .get_document(&md.table_name)
+            .map_err(Error::ValueAccess)?;
+        Ok(datasource.get(&md.col_name))
```

Comment:
TODO: need to still update the API to return `Result<Option<Bson>>` and make this returned `Ok(....cloned())`.